### PR TITLE
Add branch parameter to CST Bus Factor calculation

### PR DESCRIPTION
**Pull request by alizalisan**
This PR contains the following:
- Bus factor is calculated on the given branch(es) by the user
- If no branches are given then BF is calculated on the default branch
- Minor bug fixes

### CODE REVIEW

File: src/patterns/patterns.py [hunk 1]
```diff
@@ -722,6 +725,27 @@ def get_busfactor_data(self, locc_metric='change-size-cos', metric='mul-changes-
...
         tot_developers = 0
 
+        # picks commits from the branch(es) provided by user, else picks commits from default branch only
+        branch_df = pd.DataFrame()
+        if len(branches) == 0:
+            for b in Patterns.default_branches:
+                if len(work_df[work_df['branch'].str.contains(b)]) != 0:
+                    branches.append(b)
+                    break
+        elif len(branches) != 0:
```
Comment (fickas) [inline src/patterns/patterns.py:735]: elif => else

File: src/patterns/patterns.py [hunk 2]
```diff
@@ -722,6 +725,27 @@ def get_busfactor_data(self, locc_metric='change-size-cos', metric='mul-changes-
...
+            for i in range(len(branches)):
```
Comment (fickas) [inline src/patterns/patterns.py:736]: `for branch in branches` better?

If you need the i, you can try `for i,branch in enumerate(branches)`

**Comment by fickas** [comment]
Can you add more info on why you are adding the branches parameter? I am lacking the context.

File: src/patterns/patterns.py [hunk 3]
```diff
@@ -802,7 +802,7 @@ def get_busfactor_data(self, locc_metric='change-size-cos', metric='mul-changes-
                 authors_commits_df.iat[ind, authors_commits_df.columns.get_loc('dev_knowledge')] = d_commits/tot_commits # calculating dev_knowledge on the whole project/directory
             
             authors_commits_df.sort_values(by=['dev_knowledge'], ascending=False, inplace=True)
-            results = authors_commits_df
+            results = aggregated_df
```
Comment (fickas) [inline src/patterns/patterns.py:805]: I don't understand this change. Can you explain a bit more either in code or commit message?

File: src/patterns/patterns.py [hunk 4]
```diff
@@ -833,8 +833,8 @@ def get_busfactor_data(self, locc_metric='change-size-cos', metric='mul-changes-
                     datetime = d['datetime'][ind]
                     author = d['unique_author'][ind]
 
+            norm_factor = len(dev_knowledge_df) # total number of files in the directory, branch or project
             d = pd.DataFrame(dev_knowledge_df.groupby(['unique_author'])['dev_knowledge'].sum())
-            norm_factor = d['dev_knowledge'].sum()
             d["dev_knowledge"] = d["dev_knowledge"].apply(lambda a: a / norm_factor)
```
Comment (fickas) [inline src/patterns/patterns.py:838]: A couple comments. I'm not sure what type of scaling you are using. You appear to be dividing each value in a column by the length of the column. Is that what you want? What I was expecting to see is some type of min-max scaling where you divide every value by the max of the column. This assumes no negative values and max is not 0.

File: src/patterns/patterns.py (hunk 3 above)
Comment (alizalisan) [reply src/patterns/patterns.py:805]: I have also changed to calculating bus factor from file to folder level. Earlier I was looking at all the files in a folder but now I am aggregating like we discussed earlier. I already had that calculated in the aggregated_df, using it now.

File: src/patterns/patterns.py (hunk 2 above)
Comment (alizalisan) [reply src/patterns/patterns.py:736]: Noted.

File: src/patterns/patterns.py (hunk 1 above)
Comment (alizalisan) [reply src/patterns/patterns.py:735]: Noted.

**Comment by alizalisan** [comment]
> Can you add more info on why you are adding the branches parameter? I am lacking the context.

The branches parameter is being added because the option to look at the bus factor for one branch or multiple branches is provided originally by the authors as well. For my experiments, I am looking at the default branch only (the same happens if the branches parameter is left empty). However, leaving the parameter so that the option can be used when required.

File: src/patterns/patterns.py (hunk 3 above)
Comment (fickas) [reply src/patterns/patterns.py:805]: Still a bit confused. This code does not know seem to be needed. Am I wrong? I don't see authors_commits_df used anywhere.

            authors_commits_df["dev_knowledge"] = 0
            tot_commits = authors_commits_df[locc_metric].sum() # total number of commits on the project/directory
            for ind in authors_commits_df.index:
                d_commits = authors_commits_df[locc_metric][ind]
                authors_commits_df.iat[ind, authors_commits_df.columns.get_loc('dev_knowledge')] = d_commits/tot_commits # calculating dev_knowledge on the whole project/directory
            
            authors_commits_df.sort_values(by=['dev_knowledge'], ascending=False, inplace=True)

File: src/patterns/patterns.py (hunk 4 above)
Comment (fickas) [reply src/patterns/patterns.py:838]: I do remember discussing this but not sure I totally follow your code. But I'll give ok.

File: src/patterns/patterns.py (hunk 3 above)
Comment (alizalisan) [reply src/patterns/patterns.py:805]: Yes, that is correct. I was using it earlier but not anymore. Will remove this code.

**Review by fickas — APPROVED** [review]
Still like my for loop better on line 736.

**Comment by fickas** [comment]
Thanks for working on this.

**Comment by alizalisan** [comment]
> Still like my for loop better on line 736.

I'll change that with the next PR too. :)

**Pull request merged**